crates/matrix-sdk/src/authentication/qrcode/login.rs

```diff
@@ -0,0 +1,740 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::future::IntoFuture;
+
+use eyeball::SharedObservable;
+use futures_core::Stream;
+use mas_oidc_client::types::{
+    client_credentials::ClientCredentials, registration::VerifiedClientMetadata,
+};
+use matrix_sdk_base::{
+    boxed_into_future,
+    crypto::types::qr_login::{QrCodeData, QrCodeMode},
+    SessionMeta,
+};
+use openidconnect::DeviceCodeErrorResponseType;
+use ruma::OwnedDeviceId;
+use tracing::trace;
+use vodozemac::ecies::CheckCode;
+
+use super::{
+    messages::LoginFailureReason, oidc_client::OidcClient, DeviceAuhorizationOidcError,
+    SecureChannelError,
+};
+#[cfg(doc)]
+use crate::oidc::Oidc;
+use crate::{
+    authentication::qrcode::{
+        messages::QrAuthMessage, secure_channel::EstablishedSecureChannel, QRCodeLoginError,
+    },
+    Client,
+};
+
+async fn send_unexpected_message_error(
+    channel: &mut EstablishedSecureChannel,
+) -> Result<(), SecureChannelError> {
+    channel
+        .send_json(QrAuthMessage::LoginFailure {
+            reason: LoginFailureReason::UnexpectedMessageReceived,
+            homeserver: None,
+        })
+        .await
+}
+
+/// Type telling us about the progress of the QR code login.
+#[derive(Clone, Debug, Default)]
+pub enum LoginProgress {
+    /// We're just starting up, this is the default and initial state.
+    #[default]
+    Starting,
+    /// We have established the secure channel, but we need to let the other
+    /// side know about the [`CheckCode`] so they can verify that the secure
+    /// channel is indeed secure.
+    EstablishingSecureChannel {
+        /// The check code we need to, out of band, send to the other device.
+        check_code: CheckCode,
+    },
+    /// We're waiting for the OIDC provider to give us the access token. This
+    /// will only happen if the other device allows the OIDC provider to so.
+    WaitingForToken {
+        /// The user code the OIDC provider has given us, the OIDC provider
+        /// might ask the other device to enter this code.
+        user_code: String,
+    },
+    /// The login process has completed.
+    Done,
+}
+
+/// Named future for the [`Oidc::login_with_qr_code()`] method.
+#[derive(Debug)]
+pub struct LoginWithQrCode<'a> {
+    client: &'a Client,
+    client_metadata: VerifiedClientMetadata,
+    qr_code_data: &'a QrCodeData,
+    state: SharedObservable<LoginProgress>,
+}
+
+impl<'a> LoginWithQrCode<'a> {
+    /// Subscribe to the progress of QR code login.
+    ///
+    /// It's usually necessary to subscribe to this to let the existing device
+    /// know about the [`CheckCode`] which is used to verify that the two
+    /// devices are communicating in a secure manner.
```

Comment:
When the `Stream` ends, it's because an error happens? How do we handle errors from this `Stream`? I think it's important to explain that in the doc.

Reply:
Read the code further. Errors will be seen in `IntoFuture::into_future`. We get progress from `subscribe_to_progress` but errors from `into_future`, we need to look at 2 “channels”. Can we add a `Error(…)` variant in `LoginProgress`? It means that `into_future` should catch the errors, update the `LoginProgress` accordingly, and “rethrow” the error. That would be lovely for the consumers of `LoginProgress` as they would have only one thing to listen to.

Reply:
Hmm, that doesn't quite work with the bindings, there we have a method which returns this `Result<Client, QrLoginError>`, we would have had this function signature in the main crate but the bindings really want to test if SS is supported before logging in...

So the login had to become a two step process in the main crate. The EX guys told me that they prefer the progress to not contain the error, since they need to get the `Client` object from the future.

Perhaps we can move this function into the `ClientBuilder` and return `Result<Client, QrLoginError>` once the bindings don't need to check for SS anymore, and leave this as is for now?

Reply:
Hmm OK. Let's keep things are they are now.